libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefreshController.kt

```diff
@@ -58,41 +61,51 @@ internal class RouteRefreshController(
 
     private suspend fun tryRefreshingRoutesUntilRouteChanges(
         initialRoutes: List<NavigationRoute>
-    ): List<NavigationRoute> {
+    ): RefreshedRouteInfo {
         while (true) {
             val refreshed = refreshRoutesWithRetry(initialRoutes)
-            if (refreshed != initialRoutes) {
+            if (refreshed.routes != initialRoutes) {
                 return refreshed
             }
         }
     }
 
     private suspend fun refreshRoutesWithRetry(
         routes: List<NavigationRoute>
-    ): List<NavigationRoute> = coroutineScope {
+    ): RefreshedRouteInfo = coroutineScope {
         var timeUntilNextAttempt = async { delay(routeRefreshOptions.intervalMillis) }
         try {
             repeat(FAILED_ATTEMPTS_TO_INVALIDATE_EXPIRING_DATA) {
                 timeUntilNextAttempt.await()
                 timeUntilNextAttempt = async { delay(routeRefreshOptions.intervalMillis) }
-                val refreshedRoutes = refreshRoutesOrNull(routes)
+                val indicesSnapshot = currentIndicesSnapshotProvider.getFilledIndicesAndFreeze()
+                val refreshedRoutes = refreshRoutesOrNull(routes, indicesSnapshot.legIndex)
+                currentIndicesSnapshotProvider.unfreeze()
                 if (refreshedRoutes.any { it != null }) {
-                    return@coroutineScope refreshedRoutes.mapIndexed { index, navigationRoute ->
-                        navigationRoute ?: routes[index]
-                    }
+                    return@coroutineScope RefreshedRouteInfo(
+                        refreshedRoutes.mapIndexed { index, navigationRoute ->
+                            navigationRoute ?: routes[index]
+                        },
+                        indicesSnapshot
+                    )
                 }
             }
         } finally {
+            currentIndicesSnapshotProvider.unfreeze()
             timeUntilNextAttempt.cancel() // otherwise current coroutine will wait for its child
         }
-        routes.map { removeExpiringDataFromRoute(it) }
+        val indicesSnapshot = currentIndicesSnapshotProvider()!!
+        RefreshedRouteInfo(
+            routes.map { removeExpiringDataFromRoute(it, indicesSnapshot.legIndex) },
+            indicesSnapshot
+        )
```

Comment:
I'll take that issue right after merge. Just want to merge the PR to alpha.

And there's no need in unfreeze here. It's not frozen. It is only frozen inside try and it unfreezes in `finally` block.
And it is not frozen because we can allow updates. The only reason why we'd want to forbid them is so that `RouterWrapper` gets the same indices when it invokes the method but here `RouterWrapper` is not used.

OK, I see that it created too much of a misunderstanding, probably I should switch to mutexes now...